### PR TITLE
display of linked target field content

### LINKED ISSUE #210: only link shown for data field
When using `punx tree S-some_file.h5`, the `/entry/data/data` field is shown linked:

```
    data:NXdata
      @NX_class = "NXdata"
      @signal = "data"
      data --> b'/entry/instrument/detector/data'
```

but later, in NXinstrument, the `data` field is only described by its link and not by its content (as expected):

```
        data --> /entry/instrument/detector/data
```

Should be easy to test.

### PR DESCRIPTION
**Pull request by prjemian**
- close #210

### CODE REVIEW

**Comment by prjemian** [comment]
@carterbox Ready for review.

File: punx/h5tree.py [hunk 1]
```diff
@@ -22,6 +22,10 @@
 import logging
 import os
 import h5py
+try:  # hdf5plugin loads codecs used by h5py, don't need to call it below
+    import hdf5plugin
+except ImportError:
+    pass  # avoids unused-import report from flake8
```
Comment (carterbox) [inline punx/h5tree.py:28]: Why don't you move this to `punx/__init__.py`? Then you wouldn't need to repeat it in each module? I count this being the third time.

File: punx/h5tree.py [hunk 2]
```diff
@@ -207,15 +212,9 @@ def _renderDataset(self, dset, name, indentation="  "):
...
-                        "%s%s --> %s"
-                        % (
-                            indentation,
-                            name,
-                            utils.decode_byte_string(dset.attrs["target"]),
-                        )
-                    ]
+                target_addr = utils.decode_byte_string(dset.attrs["target"])
+                if target_addr != dset.name:
+                    return [f"{indentation}{name} --> {target_addr}"]
```
Comment (carterbox) [inline punx/h5tree.py:217]:
```suggestion
                if target_addr != dset.name:
                    return self._renderLinkedObject(
                        dset,
                        name,
                        indentation,
                    )
```

File: punx/h5tree.py [hunk 3]
```diff
@@ -320,10 +319,14 @@ def __render(obj, rank, key, indents):
...
-                part = eval("obj[%s]" % indices)
+                try:
+                    part = eval(f"obj[{indices}]")
+                except OSError as exc:
+                    return f"indices={indices}, obj={obj}, exc={exc}"
+
                 item = self._renderNdArray(part, indents + "  ")  # recursion
             else:
-                item = "rank=%d" % (rank - 1)
+                item = f"rank={rank - 1}"
```
Comment (carterbox) [inline punx/h5tree.py:329]:
```
                item = f"rank={(rank - 1):d}"
```
This is the equivalent format string, but maybe removing coercion to integer is a good idea.

File: punx/h5tree.py [hunk 4]
```diff
@@ -320,10 +319,14 @@ def __render(obj, rank, key, indents):
             elif rank < 4:
                 # this replaces a lot of code: if rank == ...
                 indices = ", ".join([str(key)] + (":" * (rank - 1)).split())
-                part = eval("obj[%s]" % indices)
+                try:
+                    part = eval(f"obj[{indices}]")
+                except OSError as exc:
```
Comment (carterbox) [inline punx/h5tree.py:324]: Curious: what causes an OSError when trying to index this object?

File: punx/tests/test_issue210.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+        # correct matches
+        "    Metadata:NXcollection",
+        "    data:NXdata",
+        '      @NX_class = "NXdata"',
+        '      @signal = "data"',
+        "      data --> /entry/instrument/detector/data",
+        '        @signal = "data"',
+        "        data:NX_UINT32[3262,3108] = __array",
+        '          @target = "/entry/instrument/detector/data"',  # FIXME
```
Comment (carterbox) [inline punx/tests/test_issue210.py:153]: data/data/data? What is happening here? Why are there so many groups called data?

**Review by carterbox — CHANGES_REQUESTED** [review]

File: punx/h5tree.py (hunk 1 above)
Comment (prjemian) [reply punx/h5tree.py:28]: Good idea

Comment (prjemian) [reply punx/h5tree.py:28]: The try..except is an attempt to quash the unused-import report from Codacy.  It does not accomplish that goal.

File: punx/h5tree.py (hunk 3 above)
Comment (prjemian) [reply punx/h5tree.py:329]: Agree.

File: punx/h5tree.py (hunk 4 above)
Comment (prjemian) [reply punx/h5tree.py:324]: An OSError is raised when the HDF5 library cannot find a compression codec it expects.

File: punx/tests/test_issue210.py (hunk 5 above)
Comment (prjemian) [reply punx/tests/test_issue210.py:153]: NeXus name soup. Default name for a group of NXthingy is `thingy`.  So NXdata group is `data` by default.  Default name for the plottable data field (Y axis, image, counts, ...) is also `data`.

Also, that `#FIXME` comment can be removed now.

File: punx/h5tree.py (hunk 4 above)
Comment (carterbox) [reply punx/h5tree.py:324]: This error is avoidable with user action (installing the codec package), so it should be accompanied by a UserWarning from the warnings module.

Comment (carterbox) [reply punx/h5tree.py:324]: Something like `UserWarning("An error occurred when reading a dataset; you probably don't have the proper HDF5 codec installed for this file.")`

File: punx/h5tree.py (hunk 3 above)
Comment (prjemian) [reply punx/h5tree.py:329]: ... that is, leaving as `item = f"rank={rank - 1}"` without the coercion

File: punx/tests/test_issue210.py (hunk 5 above)
Comment (prjemian) [reply punx/tests/test_issue210.py:153]: Succumbing to temptation, you don't imagine that's why they call it a `data` file?

File: punx/h5tree.py (hunk 1 above)
Comment (prjemian) [reply punx/h5tree.py]: Also removed the editing of `sys.path` from `punx/__init__.py`.  Was not needed.

**Comment by prjemian** [comment]
@carterbox Ready again.

File: punx/__init__.py [hunk 6]
```diff
@@ -63,15 +61,17 @@
...
-    "lxml",
-    "numpy",
-    "pyRestTable",
-    "requests",
-    # 'PyQt5',
-    # see: http://stackoverflow.com/questions/4628519/is-it-possible-to-require-pyqt-from-setuptools-setup-py
-]
+__install_requires__ = """
+    h5py
+    hdf5plugin
```
Comment (carterbox) [inline punx/__init__.py:66]: Technically not required, but as developer you can add to mandatory dependencies.

**Review by carterbox — APPROVED** [review]

**Comment by prjemian** [comment]
@carterbox Thanks for the reviews!

**Pull request merged**